File: 2022/python/aoc2022/day04.py
import sys
from typing import TextIO

from aoc2022 import Solve, run


class Solution(Solve):
    def __init__(self, io: TextIO) -> None:
        def parse(line: str) -> tuple[set[int], set[int]]:
            a0, a1 = map(lambda x: list(map(int, x.split("-"))), line.split(","))
            return (
                set(range(a0[0], a0[1] + 1)),
                set(range(a1[0], a1[1] + 1)),
            )

        self.pairs = list(map(parse, io))

    def part1(self) -> int:
        return len(list(filter(lambda a: a[0] <= a[1] or a[1] <= a[0], self.pairs)))

    def part2(self) -> int:
        return len(list(filter(lambda a: any(a[0] & a[1]), self.pairs)))


if __name__ == "__main__":
    run(Solution(sys.stdin))

File: 2025/ocaml/bin/day04.ml
open Aoc2025
open Core

module Solution : Aoc.Day = struct
  type input = (int * int) Hash_set.t

  let dirs =
    [ (-1, -1); (-1, 0); (-1, 1); (0, -1); (0, 1); (1, -1); (1, 0); (1, 1) ]

  let parse channel =
    In_channel.input_lines channel
    |> List.mapi ~f:(fun i line ->
        let f j = function
          | '@' -> Some (Tuple2.create i j)
          | '.' -> None
          | _ -> failwith "invalid input"
        in
        String.to_list line |> List.filter_mapi ~f)
    |> List.concat |> Hash_set.Poly.of_list

  let removable_rolls hs =
    Hash_set.filter hs ~f:(fun (x, y) ->
        List.map dirs ~f:(Tuple2.map2 (x, y) ~f:( + ))
        |> List.count ~f:(Hash_set.mem hs)
        < 4)

  let part1 input = removable_rolls input |> Hash_set.length |> Answer.of_int

  let part2 input =
    Sequence.unfold ~init:input ~f:(fun hs ->
        let removable = removable_rolls hs in
        match Hash_set.length removable with
        | 0 -> None
        | len -> Some (len, Hash_set.diff hs removable))
    |> Sequence.sum (module Int) ~f:Fn.id
    |> Answer.of_int
end

let () =
  let part =
    Sys.get_argv () |> Array.last |> function
    | "part1" -> Aoc.Part1
    | "part2" -> Aoc.Part2
    | _ -> Aoc.Both
  in
  Aoc.run ~part (module Solution) In_channel.stdin
